Frameworks/CoreText/DWriteWrapper_CTFont.mm

```diff
@@ -208,6 +208,173 @@ HRESULT _DWriteCreateFontFaceWithFontDescriptor(CTFontDescriptorRef fontDescript
+                                           DWRITE_INFORMATIONAL_STRING_ID id,
+                                           bool forceMatch) {
+    std::vector<ComPtr<IDWriteFont>> matchingFonts;
+    matchingFonts.reserve(fonts.size());
+    for (auto& font : fonts) {
+        BOOL exists;
+        ComPtr<IDWriteLocalizedStrings> dwriteString;
+        RETURN_IF_FAILED(font->GetInformationalStrings(id, &dwriteString, &exists));
+        if (exists) {
+            CFStringRef fontName = _CFStringFromLocalizedString(dwriteString.Get());
+            if (CFStringCompare(name, fontName, kCFCompareCaseInsensitive) == kCFCompareEqualTo) {
+                matchingFonts.emplace_back(font);
+            }
+        }
+    }
+
+    if (!matchingFonts.empty() || forceMatch) {
+        fonts = std::move(matchingFonts);
+    }
+
+    return S_OK;
+}
+
+HRESULT __FilterFontsByKeyAndValue(std::vector<ComPtr<IDWriteFont>>& fonts, CFStringRef key, CFTypeRef value, bool forceMatch) {
+    if (CFStringCompare(key, kCTFontFamilyNameAttribute, 0) == kCFCompareEqualTo) {
+        return __FilterFontsByFamilyName(fonts, static_cast<CFStringRef>(value), forceMatch);
+    } else if (CFStringCompare(key, kCTFontNameAttribute, 0) == kCFCompareEqualTo) {
+        return __FilterFontsByInformationalString(fonts,
+                                                  static_cast<CFStringRef>(value),
+                                                  DWRITE_INFORMATIONAL_STRING_POSTSCRIPT_NAME,
+                                                  forceMatch);
+    } else if (CFStringCompare(key, kCTFontDisplayNameAttribute, 0) == kCFCompareEqualTo) {
+        return __FilterFontsByInformationalString(fonts,
+                                                  static_cast<CFStringRef>(value),
+                                                  DWRITE_INFORMATIONAL_STRING_FULL_NAME,
+                                                  forceMatch);
+    } else if (CFStringCompare(key, kCTFontStyleNameAttribute, 0) == kCFCompareEqualTo) {
+        return __FilterFontsByStyleName(fonts, static_cast<CFStringRef>(value), forceMatch);
+    } else if (CFStringCompare(key, kCTFontTraitsAttribute, 0) == kCFCompareEqualTo) {
+        __FilterFontsByTraits(fonts, static_cast<CFDictionaryRef>(value), forceMatch);
+        return S_OK;
+    } else {
+        // Other attributes are currently unsupported by CTFont
+        UNIMPLEMENTED();
+        return S_FALSE;
+    }
+}
+
+#pragma endregion // Helper Filter Methods
+
+/**
+ * Creates a CFArray of CTFontDescriptors values for all fonts that match the given attributes best and completely match mandatoryKeys
+ */
+HRESULT _DWriteCreateMatchingFontDescriptors(CFDictionaryRef attributes, CFSetRef mandatoryKeys, CFArrayRef* matchingNames) {
+    *matchingNames = nullptr;
+
+    // Get list of all possible fonts
+    std::vector<ComPtr<IDWriteFont>> fonts;
+    RETURN_IF_FAILED(_DWriteGetAllFonts(fonts));
```

Comment:
ie: in the case where name or family name is specified, the 'right way' probably involves using FindFamilyName, GetFontFamily, rather than filtering ourselves. our abstraction should reflect that.

---
In reply to: [104540071](https://github.com/Microsoft/WinObjC/pull/2171#discussion_r104540071) [](ancestors = 104540071)